src/catchup/ApplyCheckpointWork.cpp

```diff
@@ -274,6 +274,8 @@ ApplyCheckpointWork::onRun()
                             lm.getLastClosedLedgerHeader())));
                 }
 
+                lm.emitAnyStoredSafeMetaAfterLedgerValidation();
```

Comment:
I think that the contract with `emitAnyStoredSafeMetaAfterLedgerValidation` is not great:
by splitting (in code) the actual work (`closeLedger`) and how we emit the meta, we provide no guarantees that we're not going to emit some random meta that was stored by a previous call to `closeLedger`.

It looks to me that the way to "idiot proof" this is to instead make it that there is a variant of `LedgerManagerImpl::closeLedger` that exists that takes the expected lcl hash as a parameter, and if present (and matches) we can emit meta right away (sharing the code path with the `UNSAFE_EMIT_META==true`).

Reply:
Oh, I see now that you'd already suggested something equivalent (I think) to a possibility I asked Graydon about above (https://github.com/stellar/stellar-core/pull/2982#discussion_r602505269) before I'd read your comment :)

Reply:
I've made this API change now (I put the new hash parameter into `LedgerCloseData`).